src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+    ):
+        """Set emulator configuration.
+
+        Args:
+            simulation_config (dict): Simulation configuration dictionary.
+            model_config (dict): Model configuration dictionary.
+            sim_opts (optional): Simulation backend specific object specifying simulation options.
+        """
+        self.simulation_config = simulation_config
+        self.model_config = model_config
+        self.sim_opts = sim_opts
+
+        self.all_simulation_backends = {"Qutip": QutipSimulator}
+
+        self.available_sweep_parameters = {
+            Parameter.amplitude,
+            Parameter.duration,
+            Parameter.frequency,
+            Parameter.relative_phase,
+            Parameter.start,
+        }
+
+        self.update()
+
+    @property
+    def sampling_rate(self):
+        return self.model_config["sampling_rate"]
+
+    def update(self):
+        """Updates the pulse simulator by loading all parameters from
+        `self.model_config` and `self.simulation_config`."""
+        self.simulation_backend_name = self.simulation_config["simulation_backend_name"]
+        self.device_name = self.model_config["device_name"]
+        self.model_name = self.model_config["model_name"]
+        self.emulator_name = f"{self.device_name} emulator running {self.model_name} on {self.simulation_backend_name} backend"
+        self.simulation_backend = self.all_simulation_backends[
+            self.simulation_backend_name
+        ](self.model_config, self.sim_opts)
+
+        self.platform2simulator_channels = self.model_config[
+            "platform2simulator_channels"
+        ]
+
+        self.sim_sampling_boost = self.simulation_config["sim_sampling_boost"]
+        self.instant_measurement = self.simulation_config["instant_measurement"]
+        self.runcard_duration_in_dt_units = self.model_config[
+            "runcard_duration_in_dt_units"
+        ]
+        # self.readout_error = self.model_config["readout_error"]
+        self.readout_error = {}
+        for k, v in self.model_config["readout_error"].items():
+            self.readout_error.update({int(k): v})
+        self.exec_params = ExecutionParameters
+        self.default_nshots = self.simulation_config["default_nshots"]
+        self.exec_params.nshots = self.default_nshots
+        self.exec_params.acquisition_type = AcquisitionType.DISCRIMINATION
+        self.simulate_dissipation = self.simulation_config["simulate_dissipation"]
+
+        self.pulse_sequence_history = []
+        self.channel_waveforms_history = []
+
+    def update_sim_opts(self, updated_sim_opts):
+        self.sim_opts = updated_sim_opts
+        self.simulation_backend.update_sim_opts(updated_sim_opts)
+
+    def connect(self):
+        log.info(f"Connecting to {self.emulator_name}.")
+
+    def disconnect(self):
+        log.info(f"Disconnecting {self.emulator_name}.")
+
+    def setup(self, *args, **kwargs):
+        log.info(f"Setting up {self.emulator_name}.")
```

Comment:
This is also not necessary to override unless you need it. It could be useful to upload any settings though (if applicable).

Reply:
It seems like I can use `setup` instead of `update` that I have, is that correct? Basically I can use it to initialise the PulseSimulator with specific settings from `simulation_config` and `sim_opts`.

Reply:
Replaced `update` with `setup` in 7f4b9230dd3d7291758bb3681abb75570ec34bf6